### PR TITLE
MARKET-1670 Image List Component

### PR DESCRIPTION
**Pull request by Fialkaksu**
The Image List displays a collection of images in an organized grid.

[MARKET-1670](https://bugs.backendless.com/browse/MARKET-1670)
[UI-Builder](https://develop.backendless.com/app/TestComponentsApp/ui-builder/ImageList/pages/landing)
[Desktop Preview](https://populardowntown.backendless.app/api/files/ui-builder/containers/ImageList/index.html?page=landing)

image

### CODE REVIEW

File: components/bl-image-list/component.json [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+      "name": "itemData",
+      "label": "Item Data",
+      "showInSettings": true,
+      "defaultValue": "[]",
+      "hasLogicHandler": true,
+      "handlerId": "itemDataLogic",
+      "handlerLabel": "Item Data Logic",
+      "dataBinding": true,
+      "settingTooltip": "Image list items data.",
+      "handlerDescription": "Specifies a JSON array containing data of the image list items. Signature of items list: [ <Item> ]. Signature of item: { img: String, title: String, rows: Number, cols: Number }."
```
Comment (serhmel) [inline components/bl-image-list/component.json:22]: Not sure about signatures in handler tooltips.

File: components/bl-image-list/src/index.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    eventHandlers.onItemClick({ event, item });
+  };
+
+  const onTitleClick = (event, title, item) => {
+    eventHandlers.onTitleClick({ event, title, item });
+  };
+
+  const onImageClick = (event, img, item) => {
+    eventHandlers.onImageClick({ event, img, item });
+  };
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:35]: use at the beginning

`  const { onItemClick, onTitleClick, onImageClick } = eventHandlers;`

and then we can remove it

File: components/bl-image-list/src/index.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+              style={ imageStyle }
+              onClick={ e => onImageClick(e, item.img, item) }/>
+
+            { showTitleBar && (
+              <ImageListItemBar
+                position={ titleBarPosition }
+                title={ item.title }
+                cols={ item.cols }
+                rows={ item.rows }
+                style={{ height:`${ titleBarHeight }px` }}
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:69]:
```suggestion
                style={{ height: `${ titleBarHeight }px` }}
```

File: components/bl-image-list/src/index.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  return useMemo(() => {
+    const heightEven = rowHeight * IMAGE_HEIGHT_REDUCTION_FACTOR;
+    const deltaHeight = rowHeight - titleBarHeight;
+    const deltaHeightEven = deltaHeight * IMAGE_HEIGHT_REDUCTION_FACTOR;
+    const isItemBarBelow = showTitleBar && titleBarPosition === POS_BELOW;
+
+    return ({
+        '--image-height': isItemBarBelow ? `${ deltaHeight }px` : `${ rowHeight }px`,
+        '--image-height-even': isItemBarBelow ? `${ deltaHeightEven }px` : `${ heightEven }px`,
+        '--image-height-common': isItemBarBelow ? `calc(100% - ${ titleBarHeight }px)` : '100%',
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:91]:
```suggestion
      '--image-height'       : isItemBarBelow ? `${ deltaHeight }px` : `${ rowHeight }px`,
      '--image-height-even'  : isItemBarBelow ? `${ deltaHeightEven }px` : `${ heightEven }px`,
      '--image-height-common': isItemBarBelow ? `calc(100% - ${ titleBarHeight }px)` : '100%',
```

File: components/bl-image-list/src/index.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+              onClick={ e => onImageClick(e, item.img, item) }/>
+
+            { showTitleBar && (
+              <ImageListItemBar
+                position={ titleBarPosition }
+                title={ item.title }
+                cols={ item.cols }
+                rows={ item.rows }
+                style={{ height:`${ titleBarHeight }px` }}
+                onClick={ e => onTitleClick(e, item.title, item) }/>
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:70]: `/>` from new line

File: components/bl-image-list/src/index.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            title={ item.title }
+            cols={ isQuilted && item.cols ? item.cols : 1 }
+            rows={ isQuilted && item.rows ? item.rows : 1 }
+            onClick={ e => onItemClick(e, item) }>
+
+            <img
+              src={ item.img }
+              alt={ item.title }
+              style={ imageStyle }
+              onClick={ e => onImageClick(e, item.img, item) }/>
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:61]: `/>` from new line

File: components/bl-image-list/src/lib/mui.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
+/**
```
Comment (serhmel) [inline components/bl-image-list/src/lib/mui.js:1]: 513,4 kB 
Not sure that all of this is required for your component

File: components/bl-image-list/src/index.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+            { showTitleBar && (
+              <ImageListItemBar
+                position={ titleBarPosition }
+                title={ item.title }
+                cols={ item.cols }
+                rows={ item.rows }
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:68]: To me it is not good idea duplicating `item` for each property.

File: components/bl-image-list/src/index.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+const { cn } = BackendlessUI.CSSUtils;
+
+const POS_BELOW = 'below';
+const TYPE_QUILTED = 'quilted';
+const IMAGE_HEIGHT_REDUCTION_FACTOR = 0.7;
+
+export default function ImageListComponent({ component, eventHandlers, elRef }) {
+  const { style, classList, display, itemData, listVariant, cols, rowHeight,
+    gap } = component;
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:13]: in single line

File: components/bl-image-list/src/index.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    return null;
+  }
+
+  return (
+    <div className={ cn('bl-customComponent-image-list', classList) } ref={ elRef } style={ style }>
+      <ImageList
+        cols={ cols }
+        gap={ gap }
+        rowHeight={ rowHeight }
+        variant={ listVariant }>
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:31]: in single line

File: components/bl-image-list/src/index.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+        { items.map(item => (
+          <Item
+            key={ item.img }
+            item={ item }
+            component={ component }
+            eventHandlers={ eventHandlers }
+          />
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:39]: in single line

File: components/bl-image-list/src/index.js [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+      cols={ isQuilted && cols ? cols : 1 }
+      rows={ isQuilted && rows ? rows : 1 }
+      onClick={ event => onItemClick({ event, item }) }>
+
+      <img
+        src={ img }
+        alt={ title }
+        style={ imageStyle }
+        onClick={ event => onImageClick({ event, img, item }) }
+      />
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:66]: in single line

File: components/bl-image-list/src/index.js [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    const heightEven = rowHeight * IMAGE_HEIGHT_REDUCTION_FACTOR;
+    const deltaHeight = rowHeight - titleBarHeight;
+    const deltaHeightEven = deltaHeight * IMAGE_HEIGHT_REDUCTION_FACTOR;
+    const isItemBarBelow = showTitleBar && titleBarPosition === POS_BELOW;
+
+    return ({
+        '--image-height'       : isItemBarBelow ? `${ deltaHeight }px` : `${ rowHeight }px`,
+        '--image-height-even'  : isItemBarBelow ? `${ deltaHeightEven }px` : `${ heightEven }px`,
+        '--image-height-common': isItemBarBelow ? `calc(100% - ${ titleBarHeight }px)` : '100%',
+    });
```
Comment (serhmel) [inline components/bl-image-list/src/index.js:93]:
```suggestion
    return ({
      '--image-height'       : isItemBarBelow ? `${ deltaHeight }px` : `${ rowHeight }px`,
      '--image-height-even'  : isItemBarBelow ? `${ deltaHeightEven }px` : `${ heightEven }px`,
      '--image-height-common': isItemBarBelow ? `calc(100% - ${ titleBarHeight }px)` : '100%',
    });
```

File: components/bl-image-list/README.md [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+<p align="center">
+  <img src="./thumbnail.png" alt="main thumbnail" width="780"/>
+</p>
+
+## Properties
+
+| Property                                   | Type                                                      | Default value | Logic           | Data Binding | UI Setting | Description                                                                                                                                                                            |
+|--------------------------------------------|-----------------------------------------------------------|---------------|-----------------|--------------|------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Item Data <br> `itemData`                  | *JSON*                                                    | `[]`          | Item Data Logic | YES          | YES        | Specifies a JSON array containing data of the image list items.  Signature of items list: `[ <Item> ]` Signature of item: `{ img: String, title: String, rows: Number, cols: Number }` |
```
Comment (v-excelsior) [inline components/bl-image-list/README.md:13]: `Item Data` -> `Items` or `Items Data`

File: components/bl-image-list/README.md [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+| Variant <br> `listVariant`                 | *Select* <br> [`standard`, `quilted`, `woven`, `masonry`] | `standard`    |                 | NO           | YES        | Controls the variant of image list to use.                                                                                                                                             |
```
Comment (v-excelsior) [inline components/bl-image-list/README.md:14]: `Variant` -> `Layout`

File: components/bl-image-list/src/index.js [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+const IMAGE_HEIGHT_REDUCTION_FACTOR = 0.7;
+
+export default function ImageListComponent({ component, eventHandlers, elRef }) {
+  const { style, classList, display, itemData, listVariant, cols, rowHeight, gap } = component;
+
+  const [items, setItems] = useState([]);
+
+  useEffect(() => {
+    setItems(itemData);
+  }, [itemData]);
```
Comment (v-excelsior) [inline components/bl-image-list/src/index.js:18]: we can avoid using of this useState and useEffect

File: components/bl-image-list/src/index.js [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+  if (!display) {
+    return null;
+  }
+
+  return (
+    <div className={ cn('bl-customComponent-image-list', classList) } ref={ elRef } style={ style }>
+      <ImageList cols={ cols } gap={ gap } rowHeight={ rowHeight } variant={ listVariant }>
+
+        { items.map(item => (
+          <Item key={ item.img } item={ item } component={ component } eventHandlers={ eventHandlers }/>
```
Comment (v-excelsior) [inline components/bl-image-list/src/index.js:29]: `item.img` is not uniq

Also, for the future(you should not change it now): `img` -> `imgURL`

Comment (v-excelsior) [inline components/bl-image-list/src/index.js:29]: Do not pass `component` here - use only needed props.

**Review by v-excelsior — CHANGES_REQUESTED** [review]

**Review by v-excelsior — APPROVED** [review]

**Pull request merged**